{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f26e733b-166c-46c0-ac6f-65d04403e94a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import mygene\n",
    "from combat.pycombat import pycombat\n",
    "import scipy.stats as stats\n",
    "import pickle\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "module_path = os.path.abspath(os.path.join('../src'))\n",
    "sys.path.insert(0, module_path)\n",
    "\n",
    "from utils import aggregate_df\n",
    "from PABCfiles import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dee2462-6108-4a15-9871-3f23bcee6937",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def readin(dfpath, columnspath, folderpath):\n",
    "# returns the df generated, and the columns generated. Made it a function cause it's the same for the pabc and nonpabc folders\n",
    "    print(\"Reading from file\")\n",
    "    df = pd.DataFrame()\n",
    "    cols = []\n",
    "    i = 1\n",
    "\n",
    "    # for each file (sample) in the folder\n",
    "    for f in os.listdir(folderpath):\n",
    "        erstatus = f.split(\",\")[3] \n",
    "        label = f.split(\",\")[0]\n",
    "\n",
    "        if os.path.isfile(os.path.join(folderpath, f)) and erstatus == \" ER pos\": # using only ER-positive tumours\n",
    "            patientfile = os.path.join(folderpath, f)\n",
    "\n",
    "            if i == 1: \n",
    "                # generating the new df based on the sample being read\n",
    "                patientdf = pd.read_csv(patientfile, skiprows=[0,1], delimiter=\"\\t\")\n",
    "                patientdf.columns = ['ID_REF', label +' VALUE']\n",
    "\n",
    "                # setting that as the growing df dataframe \n",
    "                df = patientdf.copy() \n",
    "                cols.append(label +' VALUE')\n",
    "\n",
    "            else: \n",
    "                # generating the new df based on the sample being read\n",
    "                patientdf = pd.read_csv(patientfile, skiprows=[0,1], delimiter=\"\\t\")\n",
    "                patientdf.columns = ['ID_REF', label +' VALUE']\n",
    "\n",
    "                # merging that with the growing df\n",
    "                df = df.merge(patientdf, on='ID_REF', how='inner')\n",
    "                cols.append(label +' VALUE')\n",
    "\n",
    "            i = i + 1\n",
    "    \n",
    "    file = open(dfpath,\"wb\")\n",
    "    pickle.dump(df,file)\n",
    "    file.close()\n",
    "    \n",
    "    file = open(columnspath, \"wb\")\n",
    "    pickle.dump(cols, file)\n",
    "    file.close()\n",
    "\n",
    "    return df, cols # isn't used here but could be\n",
    "    \n",
    "# loading the PABC data\n",
    "# platform used in this study: Affymetrix Human Genome U133 Plus 2.0 Array\n",
    "if not (os.path.exists(pabcpatientsdfpath) and os.path.exists(pabcpatientcolumnspath)):\n",
    "    readin(dfpath=pabcpatientsdfpath, columnspath=pabcpatientcolumnspath, folderpath=pabcpatientspath)\n",
    "\n",
    "file = open(pabcpatientsdfpath,'rb')\n",
    "pabcpatientsdf = pickle.load(file)\n",
    "file.close()\n",
    "\n",
    "file = open(pabcpatientcolumnspath,'rb')\n",
    "pabcpatientscolumns = pickle.load(file)\n",
    "file.close()\n",
    "\n",
    "# loading the nonPABC data\n",
    "# platform used in this study: Affymetrix Human Genome U133 Plus 2.0 Array (as above)\n",
    "if not (os.path.exists(nonpabcpatientsdfpath) and os.path.exists(nonpabcpatientcolumnspath)):\n",
    "    readin(dfpath=nonpabcpatientsdfpath, columnspath=nonpabcpatientcolumnspath, folderpath=nonpabcpatientspath)\n",
    "\n",
    "file = open(nonpabcpatientsdfpath,'rb')\n",
    "nonpabcpatientsdf = pickle.load(file)\n",
    "file.close()\n",
    "\n",
    "file = open(nonpabcpatientcolumnspath,'rb')\n",
    "nonpabcpatientscolumns = pickle.load(file)\n",
    "file.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "792921ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# AFFY HG U133 Plus 2 probe to ENSEMBL stable Gene id dictionary generated here: \n",
    "# http://www.ensembl.org/biomart/martview/5fee6e97d6077c05ef90f4d3b0b7b6b5\n",
    "\n",
    "affy_to_ensembl = pd.read_csv(biomartaffymetrixpath, sep='\\t').dropna()\n",
    "affy_to_ensembl.head()\n",
    "# note: there are duplicates in both the Gene stable ID column and the AFFY HG U133 Plus 2 probe column\n",
    "# this analysis converts the study's probes to Gene stable IDs (ensembl IDs) so it's okay if two\n",
    "# probes match to the same Gene stable ID, and it is likely this will happen. They are collapsed based on max value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2d4fc2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene symbol converter method that creates a dictionary from one type of gene symbol to another\n",
    "# uses DAVID gene conversion table (affymetrix_DAVID) to convert from affymetrix probe ID to ensembl ID\n",
    "# uses SynGO gene conversion table (ensembl_synGO) to convert from ensembl ID to HGNC\n",
    "\n",
    "def gene_dict(df, col1, col2):\n",
    "    d = dict()\n",
    "    for i in range(df.shape[0]):\n",
    "        if df.iloc[i][col1] not in set(d.keys()):\n",
    "            d[df.iloc[i][col1]] = [df.iloc[i][col2]]\n",
    "        else:\n",
    "            d[df.iloc[i][col1]].append(df.iloc[i][col2])\n",
    "    return d\n",
    "\n",
    "def convert(name, ref_dict):\n",
    "    if name not in set(ref_dict.keys()):\n",
    "        #print(\"NOT MATCHED: \"+ str(name))\n",
    "        return 0\n",
    "    return ref_dict[name][0] # return the first ID that was matched\n",
    "\n",
    "\n",
    "affymetrix_to_ensembl_dict = gene_dict(affy_to_ensembl, \"AFFY HG U133 Plus 2 probe\", 'Gene stable ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f7b8951",
   "metadata": {},
   "outputs": [],
   "source": [
    "nonpabcpatientsdf['Gene Symbol'] = nonpabcpatientsdf.apply(lambda row: convert(row['ID_REF'], affymetrix_to_ensembl_dict), axis=1)\n",
    "pabcpatientsdf['Gene Symbol'] = pabcpatientsdf.apply(lambda row: convert(row['ID_REF'], affymetrix_to_ensembl_dict), axis=1)\n",
    "\n",
    "nonpabcpatientsdf = nonpabcpatientsdf[nonpabcpatientsdf['Gene Symbol'] != 0]\n",
    "pabcpatientsdf = pabcpatientsdf[pabcpatientsdf['Gene Symbol'] != 0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c6f4a99",
   "metadata": {},
   "outputs": [],
   "source": [
    "class HGNCConvert:\n",
    "    def __init__(self):\n",
    "        self.mg = mygene.MyGeneInfo()\n",
    "        self.unmatched = []\n",
    "\n",
    "    def geneconv(self, g):\n",
    "        k = self.mg.getgene(g.split('.')[0], fields = 'symbol')\n",
    "        if isinstance(k, list):\n",
    "            print(g)\n",
    "            print(k)\n",
    "            self.unmatched.append(k)\n",
    "            return 0\n",
    "        if k is not None and 'symbol' in list(k.keys()):\n",
    "            return k['symbol']\n",
    "        else: \n",
    "            self.unmatched.append(k)\n",
    "            return 0\n",
    "        \n",
    "    def reset(self):\n",
    "        self.unmatched = []\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbf251ce-34ea-424c-a7ee-979891a2a33d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for each group of columns in the dictionary, perform each listed operation for each row\n",
    "# add those resulting columns to the new dataframe\n",
    "\n",
    "ensembl_to_HGNC = HGNCConvert()\n",
    "\n",
    "if not (os.path.exists(tumourdfpath)):\n",
    "    tumourdf = pd.merge(pabcpatientsdf, nonpabcpatientsdf, on = 'ID_REF', how='inner')\n",
    "    tumourdf['Ensembl ID'] = tumourdf.apply(lambda row: convert(row['ID_REF'], affymetrix_to_ensembl_dict), axis=1)\n",
    "\n",
    "    ensembl_to_symbol = {ens_id: ensembl_to_HGNC.geneconv(ens_id) for ens_id in tumourdf['Ensembl ID'].unique()}\n",
    "    tumourdf['Gene Symbol'] = tumourdf['Ensembl ID'].map(ensembl_to_symbol)\n",
    "    tumourdf = tumourdf[tumourdf['Gene Symbol'].notna()]\n",
    "    tumourdf = tumourdf[tumourdf['Ensembl ID'] != 0]\n",
    "    \n",
    "    file = open(tumourdfpath, \"wb\")\n",
    "    pickle.dump(tumourdf, file)\n",
    "    file.close()\n",
    "\n",
    "file = open(tumourdfpath,'rb')\n",
    "tumourdf = pickle.load(file)\n",
    "file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dcb91be",
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_data = tumourdf.copy().drop(columns=[\"gene_id\", \"Gene Symbol\", \"Geneid\", 'ID_REF', 'Ensembl ID'], errors='ignore')\n",
    "expression_data = expression_data[~(expression_data == 0).all(axis=1)] # drop rows with all zeros \n",
    "expression_data = expression_data[~(pd.isna(expression_data)).any(axis=1)] # drop rows with any NaN \n",
    "expression_data_T = expression_data.T\n",
    "\n",
    "\n",
    "scaler = StandardScaler()\n",
    "expression_scaled = scaler.fit_transform(expression_data_T)\n",
    "\n",
    "pca = PCA(n_components=2)  \n",
    "pca_result = pca.fit_transform(expression_scaled)\n",
    "\n",
    "\n",
    "pca_df = pd.DataFrame(pca_result, columns=['PC1', 'PC2'])\n",
    "pca_df[\"Sample\"] = expression_data_T.index\n",
    "pca_df[\"Group\"] = pca_df[\"Sample\"].apply(lambda x: 'PABC' if x in pabcpatientscolumns else \n",
    "                                                    'nonPABC' if x in nonpabcpatientscolumns else\n",
    "                                                    'UNDEFINED')\n",
    "# specific to this PABC study\n",
    "\n",
    "\n",
    "plt.figure(figsize=(10, 6))\n",
    "sns.scatterplot(x='PC1', y='PC2',\n",
    "                hue=\"Group\",  \n",
    "                data=pca_df,\n",
    "                palette=\"Set2\",  \n",
    "                )\n",
    " \n",
    "plt.title(\"PABC and nonPABC transcriptomic profiles\")\n",
    "plt.xlabel(f\"PC1 ({pca.explained_variance_ratio_[0]*100:.1f}%)\")\n",
    "plt.ylabel(f\"PC2 ({pca.explained_variance_ratio_[1]*100:.1f}%)\")\n",
    "plt.grid(True)\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84068712",
   "metadata": {},
   "outputs": [],
   "source": [
    "def conv(g):\n",
    "    k = mg.getgene(g.split('.')[0])\n",
    "    if isinstance(k, list):\n",
    "        print(g)\n",
    "        print(k)\n",
    "        return 0\n",
    "    if k is not None and 'symbol' in list(k.keys()):\n",
    "        return k['symbol']\n",
    "    else: \n",
    "        unmatched.append(k)\n",
    "        return 0\n",
    "\n",
    "if not (os.path.exists(ourdatapath)):\n",
    "    mg = mygene.MyGeneInfo()\n",
    "    unmatched = []\n",
    "    \n",
    "    ourdata = ourdataoriginal.copy()\n",
    "    ourdata['Gene Symbol'] = ourdata['Geneid'].apply(conv)\n",
    "\n",
    "    file = open(ourdatapath,'wb')\n",
    "    pickle.dump(ourdata,file)\n",
    "    file.close()\n",
    "\n",
    "\n",
    "file = open(ourdatapath,'rb')\n",
    "ourdata = pickle.load(file)\n",
    "file.close()\n",
    "\n",
    "ourdata_grouped = ourdata.copy().groupby('Gene Symbol').max()\n",
    "ourdata_grouped = ourdata_grouped.loc[~(ourdata_grouped[ourdataTPMcolumns] == 0).all(axis=1)]\n",
    "tumourdf_reset = tumourdf.copy().groupby('Gene Symbol').max()\n",
    "tumourdf_reset = tumourdf_reset.loc[~(tumourdf_reset[pabcpatientscolumns+nonpabcpatientscolumns] == 0).all(axis=1)]\n",
    "\n",
    "PABCdf = pd.merge(ourdata_grouped, tumourdf_reset, how='inner', on='Gene Symbol').reset_index()\n",
    "\n",
    "numbercols = ourdataTPMcolumns+pabcpatientscolumns+nonpabcpatientscolumns\n",
    "\n",
    "PABCdf[numbercols] = PABCdf[numbercols].apply(lambda x: np.log2(x + 1)) \n",
    "\n",
    "PABCdf # old number of rows with unique gene symbols: 18978 . after using DAVID data: 19246"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7739ecd",
   "metadata": {},
   "outputs": [],
   "source": [
    "ourdata[ourdata['Gene Symbol'].isin(k)]['D-Bewo-CT-M1'].max()\n",
    "#tumourdf_reset[tumourdf_reset['Gene Symbol'].isin(j)][pabcpatientscolumns+nonpabcpatientscolumns].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e41b222c",
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_data = PABCdf.copy().drop(columns=[\"Ensembl ID\", \"Geneid\", 'Gene Symbol', 'ID_REF'], errors='ignore')\n",
    "expression_data = expression_data[~(expression_data == 0).all(axis=1)] # drop rows with all zeros \n",
    "expression_data = expression_data[~(pd.isna(expression_data)).any(axis=1)] # drop rows with any NaN \n",
    "expression_data_T = expression_data.T\n",
    "\n",
    "\n",
    "scaler = StandardScaler()\n",
    "expression_scaled = scaler.fit_transform(expression_data_T)\n",
    "\n",
    "pca = PCA(n_components=2)  \n",
    "pca_result = pca.fit_transform(expression_scaled)\n",
    "\n",
    "\n",
    "pca_df = pd.DataFrame(pca_result, columns=['PC1', 'PC2'])\n",
    "pca_df[\"Sample\"] = expression_data_T.index\n",
    "pca_df[\"Group\"] = pca_df[\"Sample\"].apply(lambda x: 'PABC' if x in pabcpatientscolumns else \n",
    "                                                    'nonPABC' if x in nonpabcpatientscolumns else\n",
    "                                                    \"D-MCF7\" if x.split(\"-\")[0] + x.split(\"-\")[1] == \"DMCF7\" else\n",
    "                                                    \"S-MCF7\" if x.split(\"-\")[0] + x.split(\"-\")[1] == \"SMCF7\" else\n",
    "                                                    \"D-BEWO\" if x.split(\"-\")[0] + x.split(\"-\")[1] == \"DBewo\" else\n",
    "                                                    \"S-BEWO\" if x.split(\"-\")[0] + x.split(\"-\")[1] == \"SBewo\" else\n",
    "                                                    \"UNDEFINED\")\n",
    "# specific to this PABC study\n",
    "\n",
    "plt.figure(figsize=(10, 6))\n",
    "sns.scatterplot(x='PC1', y='PC2',\n",
    "                hue=\"Group\",  \n",
    "                data=pca_df,\n",
    "                palette=\"Set2\",  \n",
    "                )\n",
    " \n",
    "plt.title(\"PABC and nonPABC transcriptomic profiles\")\n",
    "plt.xlabel(f\"PC1 ({pca.explained_variance_ratio_[0]*100:.1f}%)\")\n",
    "plt.ylabel(f\"PC2 ({pca.explained_variance_ratio_[1]*100:.1f}%)\")\n",
    "plt.grid(True)\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "161740b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adapted from the pyCombat documentation:\n",
    "df_expression = PABCdf.copy()[ourdataTPMcolumns + pabcpatientscolumns + nonpabcpatientscolumns + ['Gene Symbol']]\n",
    "\n",
    "# Set index\n",
    "df_expression.set_index(['Gene Symbol'], inplace=True)\n",
    "\n",
    "# Save original index for later\n",
    "original_index = df_expression.index\n",
    "\n",
    "dataset1 = ourdata_grouped[ourdataTPMcolumns]\n",
    "dataset2 = tumourdf_reset.copy()[pabcpatientscolumns + nonpabcpatientscolumns]\n",
    "\n",
    "print(len(df_expression.columns), len(dataset1.columns), len(dataset2.columns))\n",
    "print(len(df_expression.columns) == len(dataset1.columns) + len(dataset2.columns))\n",
    "\n",
    "# we generate the list of batches\n",
    "batch = []\n",
    "datasets = [dataset1, dataset2]\n",
    "for j in range(len(datasets)):\n",
    "    batch.extend([j for _ in range(len(datasets[j].columns))])\n",
    "\n",
    "# Drop 0-variance\n",
    "zero_var_mask = df_expression.var(axis=1) == 0\n",
    "df_zero_var = df_expression[zero_var_mask]\n",
    "df_variable = df_expression[~zero_var_mask]\n",
    "\n",
    "# run pyComBat\n",
    "df_corrected_var = pycombat(df_variable,batch)\n",
    "\n",
    "df_corrected = pd.concat([df_corrected_var, df_zero_var], axis=0)\n",
    "df_corrected = df_corrected.loc[original_index]\n",
    "\n",
    "df_corrected.reset_index(inplace=True)\n",
    "df_corrected[df_corrected['D-Bewo-CT-M1'] == df_corrected['D-Bewo-CT-M1'].max()]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11bd2c0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "PABCgenes = ['HSPA5', 'BRCA1', 'BRCA2', 'SLC23A2', 'SLC23A3', 'CS', 'CPT2', 'MVK', 'NRAS','HRAS', 'RHOA', 'ROCK2', 'TGFB1', 'CDH2']\n",
    "samplesdict = {\"PABC patients\": pabcpatientscolumns,\n",
    "            #    'nonPABC patients': nonpabcpatientscolumns,\n",
    "            #    \"D-MCF7-M\":[\"D-MCF7-M1\",\"D-MCF7-M2\", \"D-MCF7-M3\"], \n",
    "            #    \"D-MCF7-S\":[\"D-MCF7-S1\", \"D-MCF7-S2\", \"D-MCF7-S3\"], \n",
    "               \"D-MCF7-CT-S\":[\"D-MCF7-CT-S1\", \"D-MCF7-CT-S2\", \"D-MCF7-CT-S3\"],\n",
    "               \"D-MCF7-CT-M\":[\"D-MCF7-CT-M1\", \"D-MCF7-CT-M2\", \"D-MCF7-CT-M3\"]}\n",
    "\n",
    "functionsdict = {\"mean\":'mean', \"STD\":np.std}\n",
    "columnstokeepdict = {\"Gene Symbol\": \"Gene Symbol\",\n",
    "                     'p-value for PABC patients vs D-MCF7-CT-M':'p-value for PABC patients vs D-MCF7-CT-M',\n",
    "                     'p-value for PABC patients vs D-MCF7-CT-S':'p-value for PABC patients vs D-MCF7-CT-S'}\n",
    "\n",
    "def significance(row, populationcol, popmeancol):\n",
    "    data = row[populationcol].astype(float).values  \n",
    "    popmean = row[popmeancol].astype(float).mean()\n",
    "    pvalue = stats.ttest_1samp(a=data, popmean=popmean)\n",
    "    return pvalue[1]\n",
    "\n",
    "Mphasecomparison = {\"patients\": pabcpatientscolumns, \"cellcondition\":[\"D-MCF7-CT-M1\", \"D-MCF7-CT-M2\", \"D-MCF7-CT-M3\"]}\n",
    "Sphasecomparison = {\"patients\": pabcpatientscolumns, \"cellcondition\":[\"D-MCF7-CT-S1\", \"D-MCF7-CT-S2\", \"D-MCF7-CT-S3\"]}\n",
    "\n",
    "PABCdflook = df_corrected[df_corrected['Gene Symbol'].isin(PABCgenes)].copy()\n",
    "\n",
    "PABCdflook['p-value for PABC patients vs D-MCF7-CT-M'] = PABCdflook.apply(lambda row: \n",
    "                                          significance(row, Mphasecomparison[\"patients\"], \n",
    "                                                            Mphasecomparison[\"cellcondition\"]), \n",
    "                                                            axis=1)\n",
    "PABCdflook['p-value for PABC patients vs D-MCF7-CT-S'] = PABCdflook.apply(lambda row: \n",
    "                                          significance(row, Sphasecomparison[\"patients\"], \n",
    "                                                            Sphasecomparison[\"cellcondition\"]), \n",
    "                                                            axis=1)\n",
    "\n",
    "aggregate_df(PABCdflook, samplesdict, functionsdict, columnstokeepdict).set_index('Gene Symbol')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59d210d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "columnstokeepdict = {'p-value for PABC patients vs D-MCF7-CT-M':'p-value for PABC patients vs D-MCF7-CT-M',\n",
    "                     'p-value for PABC patients vs D-MCF7-CT-S':'p-value for PABC patients vs D-MCF7-CT-S',\n",
    "                     }\n",
    "\n",
    "selected = df_corrected[df_corrected['Gene Symbol'].isin(PABCgenes)]\n",
    "\n",
    "PABCdf_calculated = selected.copy().groupby('Gene Symbol').max()\n",
    "\n",
    "PABCdf_calculated['p-value for PABC patients vs D-MCF7-CT-M'] = PABCdf_calculated.apply(lambda row: \n",
    "                                          significance(row, Mphasecomparison[\"patients\"], \n",
    "                                                            Mphasecomparison[\"cellcondition\"]), \n",
    "                                                            axis=1)\n",
    "PABCdf_calculated['p-value for PABC patients vs D-MCF7-CT-S'] = PABCdf_calculated.apply(lambda row: \n",
    "                                          significance(row, Sphasecomparison[\"patients\"], \n",
    "                                                            Sphasecomparison[\"cellcondition\"]), \n",
    "                                                            axis=1)\n",
    "\n",
    "\n",
    "res = aggregate_df(PABCdf_calculated, samplesdict, functionsdict, {})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e73c8e8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = res[res.index.isin(PABCgenes)].copy()\n",
    "resultspath = \"\"\n",
    "writer = pd.ExcelWriter(resultspath, engine='xlsxwriter')\n",
    "\n",
    "for gene in data.index:\n",
    "    df = data[data.index == gene]\n",
    "    df.to_excel(writer, sheet_name=gene)\n",
    "\n",
    "writer.close()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
