{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b837d02",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "module_path = os.path.abspath(os.path.join('../src'))\n",
    "sys.path.insert(0, module_path)\n",
    "\n",
    "from utils import aggregate_df, plotPCA\n",
    "\n",
    "import mygene\n",
    "from combat.pycombat import pycombat\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scipy.stats as stats\n",
    "\n",
    "import pickle\n",
    "import xlsxwriter\n",
    "\n",
    "from placentafiles import *\n",
    "ourdataoriginal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f26e733b-166c-46c0-ac6f-65d04403e94a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "studydf = pd.read_csv(studydfpath, sep='\\t')\n",
    "'''\n",
    "studydf is supplementary data from: \n",
    "Liu, Y., Fan, X., Wang, R. et al. Single-cell RNA-seq reveals the diversity of trophoblast subtypes and patterns of differentiation in the human placenta . Cell Res 28, 819–832 (2018). https://doi.org/10.1038/s41422-018-0066-y\n",
    "\n",
    "They write:\n",
    "In the present study, we isolated human villous stromal cells (STRs), CTBs, the STB, and EVTs \n",
    "during the first and second trimesters of pregnancy and monitored the transcriptome dynamics of 1567 cells at \n",
    "single-cell resolution. [...] Four populations of cells from the villi at 8 weeks of pregnancy and one \n",
    "population from the decidua at 24 weeks of pregnancy were harvested. We named these five populations of \n",
    "cells as EVT_8W (HLA-G+, mononucleated, smaller size), STB_8W (HLA-Glow, multinucleated, larger size), \n",
    "CTB_8W (CDH1+ and HLA-G-), STR_8W (HLA-G- and CDH1−), and EVT_24W (HLA-G+)\n",
    "'''\n",
    "# parse a cell sample into the cell line that was used in this study by Liu et al. (GSE89497)\n",
    "# ex: HE24W_EVT_sc1 will result in HE24W_EVT. This groups the cells into the original 5 cell lines as described in the study (GSE89497)\n",
    "def parse_placenta_cells(name):\n",
    "    ids = name.split(\"_\")\n",
    "    return (ids[0]+\"_\"+ids[1]).strip(\"1234\")\n",
    "    \n",
    "\n",
    "# create a dictionary\n",
    "# keys: 5 original cell lines\n",
    "# values: a list of all the cell samples associated with the cell line\n",
    "def group_types(lst, parse_cellLine):\n",
    "    cell_types = dict()\n",
    "    for cell in lst: \n",
    "        typ = parse_cellLine(cell)\n",
    "        if typ not in cell_types:\n",
    "            cell_types[typ] = [cell]\n",
    "        else:\n",
    "            cell_types[typ].append(cell)\n",
    "    return cell_types\n",
    "\n",
    "# the initial file from Liu et al.'s paper abbove\n",
    "studydf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9800ce2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_data = studydf.copy().drop(columns=[\"gene_id\", \"gene_name\", \"Geneid\"], errors='ignore')\n",
    "expression_data = expression_data[~(expression_data == 0).all(axis=1)] # drop rows with all zeros \n",
    "expression_data = expression_data[~(pd.isna(expression_data)).any(axis=1)] # drop rows with any NaN \n",
    "expression_data_T = expression_data.T\n",
    "\n",
    "\n",
    "scaler = StandardScaler()\n",
    "expression_scaled = scaler.fit_transform(expression_data_T)\n",
    "\n",
    "pca = PCA(n_components=2)  \n",
    "pca_result = pca.fit_transform(expression_scaled)\n",
    "\n",
    "\n",
    "pca_df = pd.DataFrame(pca_result, columns=['PC1', 'PC2'])\n",
    "pca_df[\"Sample\"] = expression_data_T.index\n",
    "pca_df[\"Group\"] = pca_df[\"Sample\"].apply(lambda x: x.split(\"_\")[0]+x.split(\"_\")[1].strip(\"1234\")) # this is the line that splits them into their cell lines\n",
    "# specific to this placenta study\n",
    "\n",
    "plt.figure(figsize=(10, 6))\n",
    "sns.scatterplot(x='PC1', y='PC2',\n",
    "                hue=\"Group\",  \n",
    "                data=pca_df,\n",
    "                palette=\"Set2\",  \n",
    "                )\n",
    " \n",
    "plt.title(\"placenta samples\")\n",
    "plt.xlabel(f\"PC1 ({pca.explained_variance_ratio_[0]*100:.1f}%)\")\n",
    "plt.ylabel(f\"PC2 ({pca.explained_variance_ratio_[1]*100:.1f}%)\")\n",
    "plt.grid(True)\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dee2462-6108-4a15-9871-3f23bcee6937",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# show the original 5 cell lines as described in the Liu et al. paper\n",
    "col_list = list(studydf)\n",
    "placentatypes = group_types(col_list, parse_placenta_cells)\n",
    "if not(os.path.exists(placentatypespath)):\n",
    "    filehandler = open(placentatypespath, 'wb')\n",
    "    pickle.dump(placentatypes, filehandler)\n",
    "    filehandler.close()\n",
    "    \n",
    "file = open(placentatypespath,'rb')\n",
    "placentatypes = pickle.load(file)\n",
    "file.close()\n",
    "    \n",
    "print(\"There are \"+ str(len(placentatypes)) + \" cell types in this dataset\")\n",
    "placentatypes.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ad7af46",
   "metadata": {},
   "outputs": [],
   "source": [
    "class HGNCConvert:\n",
    "    def __init__(self):\n",
    "        self.mg = mygene.MyGeneInfo()\n",
    "        self.unmatched = []\n",
    "\n",
    "    def geneconv(self, g):\n",
    "        k = self.mg.getgene(g.split('.')[0], fields = 'symbol')\n",
    "        if isinstance(k, list):\n",
    "            print(g)\n",
    "            print(k)\n",
    "            self.unmatched.append(k)\n",
    "            return 0\n",
    "        if k is not None and 'symbol' in list(k.keys()):\n",
    "            return k['symbol']\n",
    "        else: \n",
    "            self.unmatched.append(k)\n",
    "            return 0\n",
    "        \n",
    "    def reset(self):\n",
    "        self.unmatched = []\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40c7c082",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use mygene to transform the ensembl IDs from our data to HGNC\n",
    "# merge with the studydf\n",
    "# log transform all the data\n",
    "# placentadf is produced\n",
    "ensembl_to_HGNC = HGNCConvert()\n",
    "\n",
    "\n",
    "placentacolumns = []\n",
    "for t in ['HE24W_EVT', 'HE8W_EVT', 'HE8W_STB', 'HE8W_STR', 'HE8W_CTB']: # HE8W_CTB will be excluded later on\n",
    "    placentacolumns = placentacolumns + [c for c in placentatypes[t]]\n",
    "\n",
    "if not (os.path.exists(placentadfpath)):\n",
    "    # this includes the whole process of generating the placentadf...\n",
    "    print('generating placentadf')\n",
    "\n",
    "    ensembl_to_symbol = {ens_id: ensembl_to_HGNC.geneconv(ens_id) for ens_id in ourdataoriginal['Geneid'].unique()}\n",
    "    ourdataoriginal['Gene Symbol'] = ourdataoriginal['Geneid'].map(ensembl_to_symbol)\n",
    "    ourdataoriginal = ourdataoriginal[ourdataoriginal['Gene Symbol'].notna()]\n",
    "    studydf = studydf.reset_index().rename(columns={'index':'Gene Symbol'})\n",
    "    # takes 130 m to run\n",
    "\n",
    "    ourdata_grouped = ourdataoriginal.copy().groupby('Gene Symbol').max()\n",
    "    studydf_grouped = studydf.copy().groupby('Gene Symbol').max()\n",
    "\n",
    "    PABCdf = pd.merge(ourdata_grouped, studydf_grouped, how='inner', on='Gene Symbol').reset_index()\n",
    "\n",
    "    print('merging studydf and ourdataoriginal on Gene Symbol')\n",
    "    placentadf = pd.merge(ourdataoriginal, studydf, on='Gene Symbol', how='inner')\n",
    "\n",
    "    print('log transforming all data')\n",
    "    placentadf[placentacolumns+ourdataTPMcolumns] = placentadf[placentacolumns+ourdataTPMcolumns].apply(lambda x: np.log2(x + 1)) \n",
    "    \n",
    "    filehandler = open(placentadfpath, \"wb\")\n",
    "    pickle.dump(placentadf, filehandler)\n",
    "    filehandler.close()\n",
    "\n",
    "file = open(placentadfpath,'rb')\n",
    "placentadf = pickle.load(file)\n",
    "file.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac005579",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adapted from the pyCombat documentation:\n",
    "df_expression = placentadf.copy()\n",
    "\n",
    "# Set index\n",
    "df_expression.set_index(['Gene Symbol', 'Geneid'], inplace=True)\n",
    "# Follow same correction steps...\n",
    "# Save original index for later\n",
    "original_index = df_expression.index\n",
    "\n",
    "dataset1 = placentadf.copy()[placentacolumns]\n",
    "dataset2 = placentadf.copy()[ourdataTPMcolumns]\n",
    "\n",
    "print(len(df_expression.columns), len(dataset1.columns), len(dataset2.columns))\n",
    "print(len(df_expression.columns) == len(dataset1.columns) + len(dataset2.columns))\n",
    "\n",
    "# we generate the list of batches\n",
    "batch = []\n",
    "datasets = [dataset1, dataset2]\n",
    "for j in range(len(datasets)):\n",
    "    batch.extend([j for _ in range(len(datasets[j].columns))])\n",
    "\n",
    "# Drop 0-variance\n",
    "zero_var_mask = df_expression.var(axis=1) == 0\n",
    "df_zero_var = df_expression[zero_var_mask]\n",
    "df_variable = df_expression[~zero_var_mask]\n",
    "\n",
    "# run pyComBat\n",
    "df_corrected_var = pycombat(df_variable,batch)\n",
    "\n",
    "\n",
    "df_corrected = pd.concat([df_corrected_var, df_zero_var], axis=0)\n",
    "df_corrected = df_corrected.loc[original_index]\n",
    "\n",
    "# Add gene names back as a column \n",
    "df_corrected.reset_index(inplace=True)\n",
    "#df_corrected.rename(columns={'index': 'Gene'}, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8feeca33",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotPCA(df_corrected[ourdataTPMcolumns+placentacolumns], \"placenta transcriptome profiles VS our cell condition profiles\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbf251ce-34ea-424c-a7ee-979891a2a33d",
   "metadata": {},
   "outputs": [],
   "source": [
    "placentagenes = [\"ACVR1\",\"ADAM8\",\"ASCL2\",\"BCAM\",\"CDH1\",\"COL9A2\",\"CSH1\",\"DAG1\",\"ENG\",\"FLNA\",\n",
    "                 \"FSCN1\",\"FSTL3\",\"GATA2\",\"HGS\",\"HSPG2\",\"ITGB4\",\"JUP\",\"MMP2\",\"PAK4\",\"PIGF\",\"PKP3\",\"PLEC\",\n",
    "                 \"RRM2\",\"SLC19A1\",\"SLC1A5\",\"SLC2A8\",\"SLC38A10\",\"SLC43A1\",\"SLC43A2\",\"SLC46A1\",\"SLC7A5\",\n",
    "                 \"SNAI1\",\"VASP\",\"VAV2\",\"ZYX\"]\n",
    "\n",
    "samplesdict = {\"HE24W_EVT\": placentatypes['HE24W_EVT'],\n",
    "               \"HE8W_EVT\": placentatypes['HE8W_EVT'],\n",
    "               \"HE8W_STB\": placentatypes['HE8W_STB'],\n",
    "               \"HE8W_STR\": placentatypes['HE8W_STR'],\n",
    "               \"HE8W_CTB\": placentatypes['HE8W_CTB'],\n",
    "               \"D-Bewo-M\":[\"D-Bewo-M1\",\"D-Bewo-M2\", \"D-Bewo-M3\"], \n",
    "               \"D-Bewo-S\":[\"D-Bewo-S1\", \"D-Bewo-S2\", \"D-Bewo-S3\"], }\n",
    "\n",
    "functionsdict = {\"mean\":'mean', \"STD\":np.std}\n",
    "columnstokeepdict = {\"Gene Symbol\": \"Gene Symbol\",\n",
    "                     'p-value for HE8W_EVT vs D-Bewo-CT-M':'p-value for HE8W_EVT vs D-Bewo-CT-M',\n",
    "                     'p-value for HE8W_EVT vs D-Bewo-CT-S':'p-value for HE8W_EVT vs D-Bewo-CT-S'}\n",
    "\n",
    "def significance(row, populationcol, popmeancol):\n",
    "    data = row[populationcol].astype(float).values  \n",
    "    popmean = row[popmeancol].astype(float).mean()\n",
    "    pvalue = stats.ttest_1samp(a=data, popmean=popmean)\n",
    "    return pvalue[1]\n",
    "\n",
    "Mphasecomparison = {\"patients\": placentatypes['HE8W_EVT'], \"cellcondition\":[\"D-Bewo-CT-M1\", \"D-Bewo-CT-M2\", \"D-Bewo-CT-M3\"]}\n",
    "Sphasecomparison = {\"patients\": placentatypes['HE8W_EVT'], \"cellcondition\":[\"D-Bewo-CT-S1\", \"D-Bewo-CT-S2\", \"D-Bewo-CT-S3\"]}\n",
    "\n",
    "\n",
    "placentadflook = df_corrected[df_corrected['Gene Symbol'].isin(placentagenes)].copy().groupby('Gene Symbol').max()\n",
    "\n",
    "placentadflook['p-value for HE8W_EVT vs D-Bewo-CT-M'] = placentadflook.apply(lambda row: \n",
    "                                          significance(row, Mphasecomparison[\"patients\"], \n",
    "                                                            Mphasecomparison[\"cellcondition\"]), \n",
    "                                                            axis=1)\n",
    "placentadflook['p-value for HE8W_EVT vs D-Bewo-CT-S'] = placentadflook.apply(lambda row: \n",
    "                                          significance(row, Sphasecomparison[\"patients\"], \n",
    "                                                            Sphasecomparison[\"cellcondition\"]), \n",
    "                                                            axis=1)\n",
    "\n",
    "placentadflook = placentadflook.reset_index()\n",
    "aggregate_df(placentadflook, samplesdict, functionsdict, columnstokeepdict).set_index('Gene Symbol')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "780f5a2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = aggregate_df(placentadflook, samplesdict, functionsdict, {'Gene Symbol':'Gene Symbol'}).set_index('Gene Symbol')\n",
    "\n",
    "curresultspath = \"\"\n",
    "writer = pd.ExcelWriter(curresultspath, engine='xlsxwriter')\n",
    "\n",
    "for gene in data.index:\n",
    "    df = data[data.index == gene]\n",
    "    df.to_excel(writer, sheet_name=gene)\n",
    "\n",
    "writer.close()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
